Cython/Compiler/ExprNodes.py

```diff
@@ -11364,13 +11364,8 @@ def compute_c_result_type(self, type1, type2):
                 self, type1, type2)
 
     def py_operation_function(self, code):
-        is_unicode_concat = False
-        if isinstance(self.operand1, FormattedValueNode) or isinstance(self.operand2, FormattedValueNode):
-            is_unicode_concat = True
-        else:
-            type1, type2 = self.operand1.type, self.operand2.type
-            if type1 is unicode_type or type2 is unicode_type:
-                is_unicode_concat = type1.is_builtin_type and type2.is_builtin_type
+        type1, type2 = self.operand1.type, self.operand2.type
+        is_unicode_concat = type1 is unicode_type and type2 is unicode_type
```

Comment:
Looks like the right place to fix this, but I think we should also allow `str_type` here if one side is `unicode_type`.

Reply:
bytes + unicode doesn't work in Python - try `u"ab" + b"cd"` and `b"cd" + u"ab"`.

Reply:
However... there is one optimization it might actually be worth looking into here (although only tangentially related to this bug). CPython has a specific optimization to check if the reference count of the RHS and append rather than concat if possible: https://github.com/python/cpython/blob/309d7cc5df4e2bf3086c49eb2b1b56b929554500/Python/ceval.c#L5354

Reply:
> bytes + unicode doesn't work in Python 3

Note: I wrote `str_type`, not `bytes_type`. They are different things in Cython.